File: course-material/labs/inheritance/coding-exercises/solution/ast/addition.cpp
#include "addition.h"


Addition::Addition(const std::vector<std::shared_ptr<const Expression>>& operands)
    : Operator(operands) { }


double Addition::evaluate(const std::map<std::string, double>& context) const
{
    double result = 0;

    for (auto operand : _operands)
    {
        result += operand->evaluate(context);
    }

    return result;
}

File: course-material/labs/inheritance/coding-exercises/student/ast/java/ast/Addition.java
package ast;

import java.util.ArrayList;
import java.util.Map;

public class Addition extends Operator {
    public Addition(ArrayList<Expression> operands) {
        super(operands);
    }

    @Override
    public double evaluate(Map<String, Double> context) {
        double total = 0;

        for (Expression operand : this.operands) {
            total += operand.evaluate(context);
        }

        return total;
    }
}
